BinaryEncoding.md

```diff
@@ -3,120 +3,143 @@
-    + 64-bit offset of its operator table within the section
-  - A sequence of operator tables
-  - An operator table contains:
-    + A sequence of standardized string literal [operator names](AstSemantics.md),
-      where order determines operator index
-* A `code` section contains (in this order):
-  - The generic section header
-  - A table (sorted by offset) containing, for each function:
-     + Signature
-     + Function attributes, valid attributes TBD (could include hot/cold, optimization level, noreturn, read/write/pure, ...)
-     + 64-bit offset within the section
-  - A sequence of functions
-  - A function contains:
-    + A table containing, for each type ID that has [locals](AstSemantics.md#local-variables):
-      * Type ID
-      * Count of locals
-    + The serialized AST
-* A `data` section contains (in this order):
-   - The generic section header
-   - A sequence of byte ranges within the binary and corresponding addresses in the linear memory
-
-
-All strings are encoded as null-terminated UTF8. Data segments represent
-initialized data that is loaded directly from the binary into the linear memory
-when the program starts (see [modules](Modules.md#linear-memory-section)).
-
-## Serialized AST
-
-* Use a preorder encoding of the AST
-  * Efficient single-pass validation+compilation and polyfill
-* The data of a node (if there is any), is written immediately after the operator and before child nodes
-  * The operator statically determines what follows, so no generic metadata is necessary.
+The binary encoding is a general representation of syntax trees and module
+information that enables small files, fast decoding, and reduced memory usage.
+See the [rationale document](Rationale.md#why-a-binary-encoding) for more detail.
+
+The encoding is split into three layers:
+
+* **Layer 0** is a simple pre-order encoding of the AST and related data structures.
+  The encoding is dense and trivial to interact with, making it suitable for
+  scenarios like JIT, instrumentation tools, and debugging.
+* **Layer 1** provides structural compression on top of layer 0, exploiting
+  specific knowledge about the nature of the syntax tree and its nodes.
+  The structural compression introduces more efficient encoding of values, 
+  rearranges values within the module, and prunes structurally identical
+  tree nodes.
+* **Layer 2** applies generic compression techniques, already available
+  in browsers and other tooling. Algorithms as simple as gzip can deliver
+  good results, but more sophisticated algorithms like 
+  [LZHAM](https://github.com/richgel999/lzham_codec) and
+  [Brotli](https://datatracker.ietf.org/doc/draft-alakuijala-brotli/) are able
+  to deliver dramatically smaller files.
+
+Most importantly, the layering approach allows development and standardization to
+occur incrementally, even though production-quality implementations will need to
+implement all of the layers.
+
+# Primitives and key terminology
+
+### varuint32
```

Comment:
Probably worthwhile to go ahead and define the other integer types and to be explicit that they are encoded in little-endian.